src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+        B    4
+        Name: x, dtype: Int64
+
+        Aggregate and coerce errors to null.
+        The primary key fields are used as the index.
+
+        >>> build = builder.build(resources, errors='coerce')
+        >>> build['A']['x']
+        id
+        1       1
+        2    <NA>
+        3       4
+        Name: x, dtype: Int64
+        >>> builder.errors is None
+        True
+
+        Aggregate, log errors, and generate an error report.
+
+        >>> build = builder.build(resources, errors='log')
+        >>> build['A']['x']
+        id
+        1       1
+        2    <NA>
+        3       4
+        Name: x, dtype: Int64
+        >>> builder.errors['A']['x']
+        id
+        2    Not unique.
+        Name: x, dtype: object
+        >>> report = builder.report(build, resources)
+        >>> report['valid']
+        False
+        >>> report['resources']['A']['fields']['x']['stats']
+        {'all': 3, 'invalid': 1, 'tolerance': 0.25, 'actual': 0.333...}
+        >>> report['resources']['A']['fields']['x']['errors']
+        id
+        2    Not unique.
+        Name: x, dtype: object
+
+        Customize the error values in the error report.
+
+        >>> errorfunc = lambda x, e: as_dict(x)
+        >>> build = builder.build(resources, errors='log', errorfunc=errorfunc)
+        >>> report = builder.report(build, resources)
+        >>> report['resources']['A']['fields']['x']['errors']
+        id
+        2    {'A': [2, 2], 'B': [3]}
+        Name: x, dtype: object
+
+        Skip harvesting by turning off the behavior in the resource metadata.
+
+        >>> resources[0]['harvest'] = False
+        >>> build = builder.build(resources)
+        >>> build['A']['x']
+        id
+        1    1
+        1    1
+        2    2
+        2    2
+        Name: x, dtype: Int64
+    """
+
+    def __init__(
+        self, dfs: Union[Dict[Any, pd.DataFrame], Iterable[pd.DataFrame]]
+    ) -> None:
+        """
+        Initialize a new ResourceBuilder.
+
+        Args:
+            dfs: Either a dictionary or iterable of dataframes.
+                If an iterable, the integer index is used as dictionary keys.
+                Dataframes may not contain duplicate column basenames
+                (e.g. 'report_year', 'report_month').
+
+        Raises:
+            ValueError: Dataframe has duplicate column basenames.
+        """
+        if not isinstance(dfs, dict):
+            # Use index as integer keys
+            dfs = {i: df for i, df in enumerate(dfs)}
+        self.dfs = dfs
+        for key in self.dfs:
+            basenames = [split_period(col)[0] for col in self.dfs[key].columns]
+            if len(set(basenames)) != len(basenames):
+                raise ValueError(f"Dataframe {key} has duplicate column basenames")
+        self.errors = None
+
+    @property
+    def columns(self) -> Set[str]:
+        """Column names of all tables."""
+        return set([name for df in self.dfs.values() for name in df.columns])
```

Comment:
Right, `@property` creates a class "getter" method that is run every time the method is called (without parentheses, as if it were an instance variable). Since `self.columns` is trivial to compute, that seemed worth the cost of having it always synched with the canonical attribute `self.dfs` (i.e. robust to any changes in `self.dfs` after initialization).